[flang][driver] add -flang-deprecated-no-hlfir hidden option

jeanPerier opened the pull request:
Patch 1/3 of the transition to use the HLFIR step by default in lowering as described in https://discourse.llvm.org/t/rfc-enabling-the-hlfir-lowering-by-default/72778/7

This option will allow to lower code without the HLFIR step during a grace period as described in the RFC. It is not meant to be a long term switch for flang.

jeanPerier committed:
[flang][driver] add -flang-deprecated-no-hlfir hidden option

Patch 1/3 of the transition step 1 to HLFIR enabled by default described
in https://discourse.llvm.org/t/rfc-enabling-the-hlfir-lowering-by-default/72778/7

This option will allow to use the lowering without the HLFIR step during
a grace period until. It is not meant to be a long term switch for
flang.

tblah commented on flang/lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -1073,6 +1073,12 @@ bool CompilerInvocation::createFromArgs(
 
+  // -flang-deprecated-no-hlfir
+  if (args.hasArg(clang::driver::options::OPT_flang_deprecated_no_hlfir) &&
+      !args.hasArg(clang::driver::options::OPT_emit_hlfir)) {
```
nit: specifying -flang-depricated-nohlfir and -flang-experimental-hlfir together should probably be an error or at least a warning

tblah reviewed (changes requested):
Thanks for this!

Please could you also add testing in flang/Driver/frontend-forwarding.f90, flang/test/HLFIR/hlfir-flang.f90

jeanPerier committed:
update frontend-forwarding.f90 and add conflicting opts error

jeanPerier commented:
> Please could you also add testing in flang/Driver/frontend-forwarding.f90, flang/test/HLFIR/hlfir-flang.f90

Thanks for the review @tblah. I could not find flang/test/HLFIR/hlfir-flang.f90 test, did you mean flang/test/HLFIR/hlfir-**flags**.f90 (in which case it is updated in this PR already).

I added an error when -flang-deprecated-no-hlfir and -flang-experimental-hlfir are both specified and updated frontend-forwarding.f90.

vzakhari reviewed (approved):
Thank you, Jean!

tblah reviewed (approved):
> Thanks for the review @tblah. I could not find flang/test/HLFIR/hlfir-flang.f90 test, did you mean flang/test/HLFIR/hlfir-**flags**.f90 (in which case it is updated in this PR already).

Yes I meant hlfir-flags.f90. Sorry about that. Thanks for the changes!

jeanPerier commented:
> Yes I meant hlfir-flags.f90. Sorry about that. Thanks for the changes!
No problem, thanks for the review and comments!
